Adding an HLT filter based on PPS local track multiplicity conditions for selection of events for PPS PCL

maaraujo94 opened the pull request:
#### PR description:

This PR implements an HLT filter which selects events based on multiplicity conditions of the proton tracks from the PPS detector. It is to be used in the PCL of the PPS for the calibration streams.

#### PR validation:

Filter has been run on run 2 data and has worked as expected. More information on tests done is shown [here](https://indico.cern.ch/event/1040950/contributions/4372493/attachments/2251920/3820247/0526_PPSCal.pdf)

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

PR is not a backport.

cc: @fabferro @jjhollar @kshcheli @jan-kaspar @vavati

maaraujo94 committed:
added HLT filter for PPS calibration

maaraujo94 committed:
clang format

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34827/24606

- This PR adds an extra 16KB to repository

cmsbuild commented:
A new Pull Request was created by  @maaraujo94  for master.

It involves the following packages:

- HLTrigger/special (**hlt**)


@Martin-Grunewald, @cmsbuild can you please review it and eventually sign? Thanks.
@Martin-Grunewald this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Martin-Grunewald commented on HLTrigger/special/plugins/HLTPPSCalFilter.cc:
```diff
@@ -0,0 +1,183 @@
+
+  // strip map definition
+  // diamond map definition
+
```
Why are there no strip/diam settings?

maaraujo94 replied:
We have the strip and diam options in place but not fully defined. This is done in case there is a future decision to include these detectors in the filter as well. As they are unused for the current implementation of the filter, all references to strip / diam conditions can be removed for clarity.

Martin-Grunewald commented on HLTrigger/special/plugins/HLTPPSCalFilter.cc:
```diff
@@ -0,0 +1,183 @@
+  }
+
+  // Then filter on strips (2016-17) selection
+  if (!strip_filter_.empty()) {
```
strip is always empty as it is not being set beforehand

Martin-Grunewald commented on HLTrigger/special/plugins/HLTPPSCalFilter.cc:
```diff
@@ -0,0 +1,183 @@
+      if (strip_filter_.count(rpv.id) == 0)
+        continue;
+      auto& fltr = strip_filter_.at(rpv.id);
+      fltr = true;
```
If fltr is set to true always, then why the previous line?

maaraujo94 replied:
I am not sure what this comment is pointing out, but as mentioned in the previous reply, the strip / diam conditions can be removed, which would remove the problem for these detectors. Does this issue also apply to the case of the pixels? If so, please elaborate on what the issue is as I don't understand it at the moment.

Martin-Grunewald replied:
You set fltr to strip_filter something, and then the line immediatly afterwards you set fltr to true, so it gets reassigned a constant value...

Martin-Grunewald commented on HLTrigger/special/plugins/HLTPPSCalFilter.cc:
```diff
@@ -0,0 +1,183 @@
+  }
+
+  // Finally filter on diamond (2016+) selection
+  if (!diam_filter_.empty()) {
```
diam is always empty as it is not being set beforehand

Martin-Grunewald commented on HLTrigger/special/plugins/HLTPPSCalFilter.cc:
```diff
@@ -0,0 +1,183 @@
+      if (diam_filter_.count(rpv.id) == 0)
+        continue;
+      auto& fltr = diam_filter_.at(rpv.id);
+      fltr = true;
```
If fltr is set to true always, then why the previous line?

Martin-Grunewald reviewed (commented):
Looks like there is a lot of unneeded code included, or some other code is missing...

Martin-Grunewald commented on HLTrigger/special/plugins/HLTPPSCalFilter.cc:
```diff
@@ -0,0 +1,183 @@
+  // maps for assigning filter pass / fail
+  std::unordered_map<uint32_t, bool> pixel_filter_;
+  std::unordered_map<uint32_t, bool> strip_filter_;
+  std::unordered_map<uint32_t, bool> diam_filter_;
```
It seems these 3 maps are just used as bools to switch on/off IDs, so these variables could become (const) class members and initialised as such in the class, rather than being local variables.

maaraujo94 replied:
The maps are used as bools to store whether the track multiplicity of each detector is within the max and min number of tracks. The final filter condition combines the result for each detector in a single logical expression, depending on which type of filtering is done. 
Given their purpose, I'm not sure how to define the maps as const class members and initialize them in the class, and still be able to change their value when the filter checks track multiplicity conditions. Could you give an example of how to implement this?

qliphy commented:
@maaraujo94  Would you please address above comments?

maaraujo94 commented:
Thank you for the reminder. I have addressed the comments on the code and will await responses to these, then implement the necessary changes and push the updates.

Martin-Grunewald commented:
Please remove placeholder/unused code.

Martin-Grunewald commented on HLTrigger/special/plugins/HLTPPSCalFilter.cc:
```diff
@@ -0,0 +1,183 @@
+  pixel_filter_[CTPPSPixelDetId(0, 2, 3)] = false;
+  pixel_filter_[CTPPSPixelDetId(1, 0, 3)] = false;
+  pixel_filter_[CTPPSPixelDetId(1, 2, 3)] = false;
+
```
pixel_filter never gets changed after these lines, so can be class members initialised in the class.
Same for the other two Boolean arrays.

maaraujo94 committed:
removed unused code; removed auxiliary variable fltr

maaraujo94 commented:
I have updated and pushed the code, removing the unused code for strips and diam and also removing the auxiliary variable "fltr", which is not essential and which I believe was obscuring how the filter is processed, raising issues about the definitions of the maps of booleans.
Please let me know if I have misunderstood any of the issues above and they still stand in the current form of the filter.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34827/24748

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #34827 was updated. @Martin-Grunewald, @cmsbuild can you please check and sign again.

Martin-Grunewald commented:
I still think the Boolean unordred_maps should be class members and initialised in the constructor (even if not const, syntax becomes ugly), as the maps do not seem to be changed during the event processing.

maaraujo94 commented:
I have attempted this change and found that, due to the definition of the filter as "bool filter (...) const", it is not possible to alter the boolean values of the map inside the filter if they are class members (lines 98, 104, 106 of the code give compilation errors related to this). This const declaration does not seem to be removable as it's an imposition for a global EDFilter. Please advise on which changes to apply to the filter declaration and class type to circumvent this issue.

Martin-Grunewald commented:
Ah ok, never mind then.

Martin-Grunewald commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f2761f/17909/summary.html
**COMMIT**: aee5ef297db1fdd7d69de191c9445bb731c6cc02
**CMSSW**: CMSSW_12_1_X_2021-08-19-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/34827/17909/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 3 differences found in the comparisons
 - DQMHistoTests: Total files compared: 39
 - DQMHistoTests: Total histograms compared: 3000352
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3000324
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 38 files compared)
 - Checked 165 log files, 37 edm output root files, 39 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-08-19-2300+f2761f/44890/triggerResults): no differences found

Martin-Grunewald commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented on HLTrigger/special/plugins/HLTPPSCalFilter.cc:
```diff
@@ -0,0 +1,124 @@
+  bool filter(edm::StreamID, edm::Event&, const edm::EventSetup&) const override;
+
+private:
+  edm::InputTag pixelLocalTrackInputTag_;  // Input tag identifying the pixel detector
```
All these private class members can be declared `const`

perrotta commented on HLTrigger/special/plugins/HLTPPSCalFilter.cc:
```diff
@@ -0,0 +1,124 @@
+HLTPPSCalFilter::HLTPPSCalFilter(const edm::ParameterSet& iConfig)
+    : pixelLocalTrackInputTag_(iConfig.getParameter<edm::InputTag>("pixelLocalTrackInputTag")),
+      pixelLocalTrackToken_(consumes<edm::DetSetVector<CTPPSPixelLocalTrack>>(pixelLocalTrackInputTag_)),
+
```
Removing this empty line can improve readability

perrotta commented on HLTrigger/special/plugins/HLTPPSCalFilter.cc:
```diff
@@ -0,0 +1,124 @@
+
+      minTracks_(iConfig.getParameter<int>("minTracks")),
+      maxTracks_(iConfig.getParameter<int>("maxTracks")),
+
```
Removing this empty line can improve readability

perrotta commented on HLTrigger/special/plugins/HLTPPSCalFilter.cc:
```diff
@@ -0,0 +1,124 @@
+  pixel_filter_[CTPPSPixelDetId(1, 2, 3)] = false;
+
+  // filter on pixels (2017+) selection
+  if (!pixel_filter_.empty()) {
```
Can `pixel_filter_` still be empty here? It is filled right above

maaraujo94 committed:
applied suggested changes

maaraujo94 commented:
I have applied the suggested changes to the code, including removing the check "if (!pixel_filter_.empty())" which is redundant

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34827/24846

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #34827 was updated. @Martin-Grunewald, @cmsbuild can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34827/24847

- This PR adds an extra 16KB to repository

perrotta commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f2761f/17994/summary.html
**COMMIT**: 69e510a3156e69bcb950010e0fafa9b0f81e29aa
**CMSSW**: CMSSW_12_1_X_2021-08-23-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/34827/17994/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 3 differences found in the comparisons
 - DQMHistoTests: Total files compared: 39
 - DQMHistoTests: Total histograms compared: 3000352
 - DQMHistoTests: Total failures: 5
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 3000324
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.004 KiB( 38 files compared)
 - DQMHistoSizes: changed ( 312.0 ): 0.004 KiB        MessageLogger/Warnings
 - Checked 165 log files, 37 edm output root files, 39 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-08-23-2300+f2761f/44962/triggerResults): no differences found

Martin-Grunewald commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
+1
